### PR TITLE
Derive Error implementation

### PR DESCRIPTION
**Pull request by wellsie1116**
Replace the manual implementation of std::error::Error for Error with the thiserror::Error derive macro. Other than the error strings containing more information and thiserror::Error implementing std::error::Error::description() with a static string indicating the method is deprecated, no observable changes expected.

Related to #136

### CODE REVIEW

**Comment by heartsucker** [comment]
I'm not sure how this is more useful than `failure` or why this should be picked. I have used `failure` in other places and been happy with the way that API works.. I also don't see much of a change with this compared to the current impl. What's the benefit here?

**Comment by wellsie1116** [comment]
First off, I am 100% okay with using failure.  I too have used it and like its API surface and derive macro.

thiserror/anyhow are very similar, except:
* they are based on the std::error::Error trait, which is less broken these days, removing some of the incentive to depend on or export a custom trait
* they are at 1.0, failure has no plans to stabilize
* they produce nicer error output in my opinion
* thiserror allows fields to be tagged with #[from] to auto-generate From<> conversion to that error variant

This pull request as-is doesn't change much.  The Display implementation will include the context from the error variants now and the "soft-deprecated" display method is now a stub, but that is about it.

The reason why I'm poking at the `Error` type at all is that I am interested in teasing apart some of the cases in Error::Opaque.  Specifically, I'd like to know if a Client::update failed due to a network related error.  But if I have an error variant that wraps a `hyper::Error`, then I should probably implement `std::error::Error::source`, and if I'm going to implement `source`, I might as well switch to a derive helper, so here we are.

So, my understanding of thiserror/anyhow is it does roughly everything failure does in a subjectively prettier and more stable way, but again, I'm fine with using either.

Also, hi :)

**Review by erickt — APPROVED** [review]
I like this too, but I'll wait for @heartsucker to approve before merging since they haven't responded to your response yet.

**Comment by heartsucker** [comment]
I think this is reasonable, and at the time I wrote this, I didn't have a way to nicely include more detailed context in the `Opaque`. I think error handling in this could be overall improved, and I know that `failure` has some modes of usage that allow for boxing and downcasting as "pattern matching" which would allow for some extraction.

I agree that this is an improvement, and in theory we could improve the overall structure under here too.

**Pull request merged**
